src/client/VZCodeContext.tsx

```diff
@@ -172,6 +178,52 @@ export const VZCodeProvider = ({
     setUsername,
   } = useActions(dispatch);
 
+  // initialize open tabs from the url search parameters
+  useEffect(() => {
+    const search = new URLSearchParams(
```

Comment:
Very nice! I would prefer to use React Router, so that we can leverage their various utilities.